skimage/filters/_multiotsu.pyx

```diff
@@ -6,101 +6,426 @@ import numpy as np
+        if zeroth_moment[i] > 0:
+            var_btwcls[i] = (first_moment[i]**2) / zeroth_moment[i]
+
+    idx = nbins
+
+    for i in range(1, nbins):
+        for j in range(i, nbins):
+            zeroth_moment_ij = zeroth_moment[j] - zeroth_moment[i - 1]
+            if zeroth_moment_ij > 0:
+                first_moment_ij = first_moment[j] - first_moment[i - 1]
+                var_btwcls[idx] = (first_moment_ij**2) / zeroth_moment_ij
+            idx += 1
+
+
+cdef float _get_var_btwclas_lut(float [::1] var_btwcls, Py_ssize_t i,
+                                Py_ssize_t j, Py_ssize_t nbins) nogil:
+    """Returns the between classes variance stored in compressed upper
+    triangular matrix form at the desired 2D indices.
+
+    Parameters
+    ----------
+    var_btwcls: 1D array
+        The between classes variance lookup table in compressed upper
+        triangular matrix form.
+    i, j: int
+        2D indices in the uncompressed lookup table.
+    nbins: int
+        The number of columns in the lookup table.
 
+    Returns
+    -------
+    value: float
+        The value of the lookup table corresponding to index (i, j).
 
-cdef double _find_best_rec(double[:, ::1] var_btwcls, cnp.intp_t min_val,
-                           cnp.intp_t max_val, Py_ssize_t[::1] idx_tuple,
-                           cnp.intp_t divisions, cnp.intp_t depth,
-                           double max_sigma, Py_ssize_t[::1] aux_thresh) nogil:
     """
-    Recursive function for calculating max_sigma.
+    cdef cnp.intp_t idx = (i * (2 * nbins - i + 1)) / 2 + j - i
+    return var_btwcls[idx]
+
+
+cdef float _set_thresh_indices_lut(float[::1] var_btwcls, Py_ssize_t hist_idx,
+                                   Py_ssize_t thresh_idx, Py_ssize_t nbins,
+                                   Py_ssize_t thresh_count, float sigma_max,
+                                   Py_ssize_t[::1] current_indices,
+                                   Py_ssize_t[::1] thresh_indices) nogil:
+    """Recursive function for finding the indices of the thresholds
+    maximizing the  variance between classes sigma.
+
+    This implementation use a lookup table of between classes variance
+    to perform a brute force evaluation of sigma over all the
+    combinations of threshold to find the indices maximizing sigma
+    (see [1]_).
 
     Parameters
     ----------
-    var_btwcls : 2-d array
-        Array of variance between classes.
-    min_val : int
-        Minimum value of the checked intervals.
-    max_val : int
-        Maximum value of the checked intervals.
-    idx_tuple : array
+    var_btwcls: (nbins*(nbins + 1) /2) 1D array
+        The Upper triangular part of the between classes variance
+        lookup table.
+    hist_idx : int
+        Current index in the histogram.
+    thresh_idx : int
+        Current index in thresh_indices.
+    nbins : int
         number of bins used in the histogram
-    divisions : int
+    thresh_count : int
```

Comment:
Now that I'm seeing what this variable represents, its name is not clear at all... couldn't it be something like `classes_count`, or `classes_gen`, according to the documentation?

Reply:
Here `thresh_count` is really the number of returned threshold values, so it is the number of classes  - 1.

Reply:
Could you fix the documentation for this variable, then? It seems we're talking classes, not thresholds.

Reply:
It's fixed ;-) thank you again for your review.